core/src/main/java/com/expedia/adaptivealerting/core/detector/PewmaOutlierDetector.java

```diff
@@ -0,0 +1,186 @@
+
+
+    /**
+     * Outlier weighting param.
+     */
+    private final double beta;
+
+    /**
+     * How many iterations to train for.
+     */
+    private final double trainingLength;
+
+    /**
+     * Smoothing param.
+     */
+    private double trainingCount;
+
+    /**
+     * Internal state for PEWMA algorithm.
+     */
+    private double s1;
+
+    /**
+     * Internal state for PEWMA algorithm.
+     */
+    private double s2;
+
+    /**
+     * Weak outlier threshold, in sigmas.
+     */
+    private final double weakThreshold;
+
+    /**
+     * Strong outlier threshold, in sigmas.
+     */
+    private final double strongThreshold;
+
+    /**
+     * Local mean estimate.
+     */
+    private double mean;
+
+    /**
+     * Local standard deviation estimate.
+     */
+    private double stdDev;
+
+    /**
+     * Creates a new PEWMA anomaly detector. Initial mean is given by initValue and initial standard deviation is 0.
+     *
+     * @param initialAlpha    Smoothing parameter.
+     * @param beta            Outlier weighting parameter.
+     * @param trainingLength  How many iterations to train for.
+     * @param weakThreshold   Weak outlier threshold, in sigmas.
+     * @param strongThreshold Strong outlier threshold, in sigmas.
+     * @param initValue       Initial observation, used to set the first mean estimate.
+     */
+    public PewmaOutlierDetector(
+            double initialAlpha,
+            double beta,
+            int trainingLength,
+            double weakThreshold,
+            double strongThreshold,
+            double initValue
+    ) {
+        isBetweenExclusive(initialAlpha, 0.0, 1.0, "initialAlpha must be in the range (0, 1)");
+
+        this.alpha0 = 1 - initialAlpha;  // To standardise with the EWMA implementation we use the complement value.
+        this.beta = beta;
+        this.trainingLength = trainingLength;
+        this.trainingCount = 1;
+        this.weakThreshold = weakThreshold;
+        this.strongThreshold = strongThreshold;
+        this.s1 = initValue;
+        this.s2 = pow(initValue, 2);
+        updateMeanAndStdDev();
+    }
+
+    private void updateMeanAndStdDev() {
+        this.mean = this.s1;
+        this.stdDev = sqrt(this.s2 - pow(this.s1, 2));
+    }
+    
+    @Override
+    public OutlierLevel evaluate(Instant instant, double value) {
+        
+        // TODO Currently we just ignore the instant. Would be helpful to have a way to detect missing data points in
+        // the stream, and to define strategies for dealing with this. [WLW]
+
+        final double dist = abs(value - mean);
+        final double largeThreshold = strongThreshold * stdDev;
+        final double smallThreshold = weakThreshold * stdDev;
+
+        OutlierLevel level = OutlierLevel.NORMAL;
+        if (dist > largeThreshold) {
+            level = OutlierLevel.STRONG;
+        } else if (dist > smallThreshold) {
+            level = OutlierLevel.WEAK;
+        }
```

Comment:
Very similar logic in `EwmaOutlierDetector`. Might need to think about combining somehow.

Reply:
Happy to wait for the patterns to emerge.